src/TesApi.Web/Storage/TerraStorageAccessProvider.cs

```diff
@@ -0,0 +1,161 @@
+    public class TerraStorageAccessProvider : StorageAccessProvider
+    {
+        private readonly TerraOptions terraOptions;
+        private readonly TerraWsmApiClient terraWsmApiClient;
+        private const string SasBlobPermissions = "racw";
+        private const string SasContainerPermissions = "racwl";
+
+        /// <summary>
+        /// Provides methods for blob storage access by using local path references in form of /storageaccount/container/blobpath
+        /// for Terra
+        /// </summary>
+        /// <param name="logger">Logger <see cref="ILogger"/></param>
+        /// <param name="terraOptions"><see cref="TerraOptions"/></param>
+        /// <param name="azureProxy">Azure proxy <see cref="IAzureProxy"/></param>
+        /// <param name="terraWsmApiClient"><see cref="terraWsmApiClient"/></param>
+        public TerraStorageAccessProvider(ILogger<TerraStorageAccessProvider> logger,
+            IOptions<TerraOptions> terraOptions, IAzureProxy azureProxy, TerraWsmApiClient terraWsmApiClient) : base(
+            logger, azureProxy)
+        {
+            this.terraWsmApiClient = terraWsmApiClient;
+            ArgumentNullException.ThrowIfNull(terraOptions);
+
+            this.terraOptions = terraOptions.Value;
+        }
+
+        /// <inheritdoc />
+        public override Task<bool> IsPublicHttpUrlAsync(string uriString)
+        {
+            var isHttpUrl = TryParseHttpUrlFromInput(uriString, out var uri);
+
+            if (!isHttpUrl)
+            {
+                return Task.FromResult(false);
+            }
+
+            if (HttpUtility.ParseQueryString(uri.Query).Get("sig") is not null)
+            {
+                return Task.FromResult(true);
+            }
+
+            if (StorageAccountUrlSegments.TryCreate(uriString, out var parts))
+            {
+                if (IsTerraWorkspaceContainer(parts.ContainerName) && IsTerraWorkspaceStorageAccount(parts.AccountName))
+                {
+                    return Task.FromResult(false);
+                }
+            }
+
+            return Task.FromResult(true);
+        }
+
+        /// <inheritdoc />
+        public override async Task<string> MapLocalPathToSasUrlAsync(string path, bool getContainerSas = false)
+        {
+            ArgumentException.ThrowIfNullOrEmpty(path);
+
+            var normalizedPath = path.TrimStart('/');
+
+            if (getContainerSas)
+            {
+                return await GetMappedSasContainerUrlFromWsmAsync(normalizedPath);
+            }
+
+            if (IsItKnownExecutionFilePath(normalizedPath))
+            {
+                return await GetMappedSasUrlFromWsmAsync(normalizedPath);
+            }
+
+            if (!StorageAccountUrlSegments.TryCreate(normalizedPath, out var segments))
+            {
+                throw new Exception(
+                    "Invalid path provided. The path must be a valid blob storage url or a path with the following format: /accountName/container");
```

Comment:
Can the code ever get here if the URL is a random Internet file?  I assume not...

Reply:
This code was updated.